libdrgn/iterators.c

```diff
@@ -0,0 +1,472 @@
+	if (err)
+		goto out;
+	err = drgn_object_set_unsigned(&iter->pos, void_star_type, 0, 0);
+	if (err)
+		goto out;
+	err = drgn_object_copy(&iter->ns, ns);
+	if (err)
+		goto out;
+	err = drgn_program_find_type(prog, "struct pid", NULL, &iter->pid_type);
+	if (err)
+		goto out;
+	err = drgn_program_find_type(prog, "struct upid", NULL,
+				     &iter->upid_type);
+	if (err)
+		goto out;
+	err = drgn_program_find_object(prog, "pidhash_shift", NULL,
+				       DRGN_FIND_OBJECT_ANY, &pidhash_shift);
+	if (err)
+		goto out;
+	err = drgn_object_read_unsigned(&pidhash_shift, &iter->hash_length);
+	if (err)
+		goto out;
+	iter->hash_length = 1 << iter->hash_length;
+	err = drgn_object_member_dereference(&ns_level, ns, "level");
+	if (err)
+		goto out;
+	uint64_t ns_level_value;
+	err = drgn_object_read_unsigned(&ns_level, &ns_level_value);
+	if (err)
+		goto out;
+	snprintf(iter->member_specifier, sizeof(iter->member_specifier),
+		 "numbers[%" PRIu64 "]", ns_level_value);
+out:
+	if (err) {
+		drgn_object_deinit(&iter->pid_hash);
+		drgn_object_deinit(&iter->pos);
+		drgn_object_deinit(&iter->ns);
+		drgn_object_deinit(&iter->entry);
+	}
+	drgn_object_deinit(&pidhash_shift);
+	drgn_object_deinit(&ns_level);
+	return err;
+}
+
+struct drgn_error *pid_iter_init(struct pid_iter *iter,
+				 const struct drgn_object *ns)
+{
+	struct drgn_program *prog = drgn_object_program(ns);
+	struct drgn_error *err;
+	struct drgn_object idr;
+	drgn_object_init(&idr, prog);
+
+	err = drgn_object_member_dereference(&idr, ns, "idr");
+	if (!err) {
+		iter->has_idr = true;
+		err = drgn_program_find_type(prog, "struct pid *", NULL,
+					     &iter->pid_type);
+		if (err)
+			goto out;
+		err = idr_iter_init(&iter->iter, &idr);
+	} else if (err->code == DRGN_ERROR_LOOKUP) {
+		drgn_error_destroy(err);
+		err = pid_iter_init_old_kernel(prog, ns, iter);
+	}
+
+out:
+	drgn_object_deinit(&idr);
+	return err;
+}
+
+void pid_iter_deinit(struct pid_iter *iter)
+{
+	if (iter->has_idr) {
+		idr_iter_deinit(&iter->iter);
+	} else {
+		drgn_object_deinit(&iter->pid_hash);
+		drgn_object_deinit(&iter->pos);
+		drgn_object_deinit(&iter->ns);
+		drgn_object_deinit(&iter->entry);
+	}
+}
+
+struct drgn_error *pid_iter_next(struct pid_iter *iter,
+				 struct drgn_object **ret)
+{
+	struct drgn_error *err;
+	if (iter->has_idr) {
+		struct radix_tree_iter_entry *entry;
+		err = idr_iter_next(&iter->iter, &entry);
+		if (err)
+			return err;
+		if (!entry) {
+			*ret = NULL;
+			return NULL;
+		}
+		err = drgn_object_cast(&entry->node, iter->pid_type,
+				       &entry->node);
+		if (!err)
+			*ret = &entry->node;
+		return err;
+	} else {
+		struct drgn_object upid;
+		struct drgn_object upid_ns;
+		bool done = false;
+		drgn_object_init(&upid, drgn_object_program(&iter->ns));
+		drgn_object_init(&upid_ns, drgn_object_program(&iter->ns));
+
+		while (!done &&
+		       (iter->index == -1 || iter->index < iter->hash_length)) {
```

Comment:
This logic can be simplified by doing something along these lines:

```python
class _PidHashIterator:
    def __init__(self, ns: Object) -> None:
        self.pid_hash = ns.prog_["pid_hash"]
        self.member = f"numbers[{int(ns.level)}]"
        self.ns = ns
        self.i = 1 << ns.prog_["pidhash_shift"]
        self.node = Object(ns.prog_, "void *", 0)

    def __iter__(self):
        return self

    def __next__(self) -> Object:
        while True:
            while not self.node:
                if self.i == 0:
                    raise StopIteration
                self.i -= 1
                self.node = self.pid_hash[self.i].first
            while self.node:
                upid = container_of(self.node, "struct upid", "pid_chain")
                self.node = self.node.next
                if upid.ns == self.ns:
                    return container_of(upid, "struct pid", self.member)
```

Namely:
* Iterate backwards so you don't need to save an index and the hash table length.
* Handle finding the next non-empty bucket separately from finding a matching entry in the bucket, which reduces the number of special cases for starting the iteration.

Reply:
Wouldn't this return the pids out of order though? I realize that the pids aren't guaranteed to be strictly monotonically increasing, but they are nearly sorted when you iterate forwards. Is there no added value in this?

Reply:
The current implementation on Linux 4.14 gave me

```
>>> [pid.numbers[0].nr.value_() for pid in for_each_pid(prog)]
[131, 42, 8, 63, 29, 50, 16, 71, 37, 3, 24, 134, 45, 11, 66, 32, 53, 19, 129, 40, 6, 61, 27, 48, 14, 69, 35, 1, 22, 132, 43, 9, 64, 30, 51, 17, 38, 93, 4, 25, 135, 46, 12, 67, 33, 54, 109, 20, 130, 41, 7, 62, 28, 49, 15, 70, 36, 2, 23, 133, 44, 10, 65, 31, 52, 18, 39, 5, 26, 136, 47, 13, 68, 34, 55, 21]
```

Looking at the hash function (see [here](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/kernel/pid.c?h=v4.14.258#n43) and [here](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/include/linux/hash.h?h=v4.14.258)), it seems to mix the PID entirely. So, I think the hash table is in a random order and thus it doesn't make a difference which direction we iterate.

(For the radix tree implementation, yes, that's in sorted order.)